src/Language/DifferentialDatalog/CompileJava.hs

```diff
@@ -0,0 +1,487 @@
+-}
+
+module Language.DifferentialDatalog.CompileJava (
+    compileJava,
+) where
+
+import Prelude hiding((<>))
+import Control.Monad.State
+import Text.PrettyPrint
+import Data.Tuple
+import Data.Tuple.Select
+import Data.Maybe
+import Data.List
+import Data.Int
+import Data.Word
+import Data.Bits hiding (isSigned)
+import Data.List.Split
+import Data.FileEmbed
+import Data.String.Utils
+import Data.Char
+import System.FilePath
+import System.Directory
+import qualified Data.ByteString.Char8 as BS
+import Numeric
+import qualified Data.Set as S
+import qualified Data.Map as M
+import qualified Data.Graph.Inductive as G
+import qualified Data.Graph.Inductive.Query.DFS as G
+import Debug.Trace
+import Text.RawString.QQ
+import Data.WideWord
+import Debug.Trace
+
+import Language.DifferentialDatalog.PP
+import Language.DifferentialDatalog.Name
+import Language.DifferentialDatalog.Pos
+import Language.DifferentialDatalog.Ops
+import Language.DifferentialDatalog.Util
+import Language.DifferentialDatalog.Syntax
+import Language.DifferentialDatalog.Parse
+import Language.DifferentialDatalog.NS
+import Language.DifferentialDatalog.Expr
+import Language.DifferentialDatalog.DatalogProgram
+import Language.DifferentialDatalog.Relation
+import Language.DifferentialDatalog.Optimize
+import Language.DifferentialDatalog.Module
+import Language.DifferentialDatalog.ECtx
+import Language.DifferentialDatalog.Type
+import Language.DifferentialDatalog.Rule
+
+-- generate a Java program from a DDlog program
+-- The Java program has some utility function to more easily create
+-- records for the DDlog relations.  Note that Java does not support
+-- all DDlog constructs
+compileJava :: DatalogProgram -> String -> IO ()
+compileJava d sourceName = do
+    let content = generateJava d sourceName in
+         writeFile (capitalize sourceName ++ ".java") (render content)
+
+-- Generate a Java document from a program given the source filename
+generateJava :: DatalogProgram -> String -> Doc
+generateJava d sourceName =
+    preamble sourceName $+$
+    (braces' $ body d)
+
+-- Generates the preamble of a Java file
+preamble :: String -> Doc
+preamble sourceName =
+    "import java.util.*;"                       $+$
+    "import ddlogapi.DDlogRecord;\n"            $+$
+    "import ddlogapi.DDlogCommand;\n"           $+$
+    "class" <+> pp (capitalize sourceName)
+
+data JavaState = JavaState {
+    -- types that have been compiled
+    typesCompiled  :: S.Set String,
+    -- types that still have to be compiled
+    typesToCompile :: M.Map String Type
+}
+
+type JavaMonad = State JavaState
+emptyState :: JavaState
+emptyState = JavaState {
+    typesCompiled  = S.empty,
+    typesToCompile = M.empty
+}
+
+-- Generates the body of the Java class containing all utility methods
+body :: DatalogProgram -> Doc
+body d =
+    evalState (
+        do relations <- mapM (genRelationCreate d) (M.keys $ progRelations d)
+           -- TODO: this should iterate until todo is empty
```

Comment:
The simplest way to do this is to add a function to `Type.hs` to recursively compute all types mentioned in a type declaration (I can do this), call this function for each type in `typesToCompile`, and union the results. Then you won't need fix point here, and `genCreate` won't have to return a monad.

Reply:
I thought about this; we also need to be able to tell whether a type comes from a relation or not, since we generate slightly different code.